console: refactor inspector console extension installation

joyeecheung opened the pull request:
- Instead of creating the console extensions eagerly during bootstrap
  and storing them on an object, wrap the code into a function to be
  called during `installAdditionalCommandLineAPI` only when the extensions
  are actually needed, which may not even happen if the user do not
  use the console in an inspector session, and does not need to happen
  during bootstrap unconditionally.
- Port `internalBinding('inspector').consoleCall` to JS, so that we do
  not need to call twice from C++ into JS for each console method call
  and the code is easier to understand. We can simply query if the
  inspector is active with `internalBinding('inspector').isEnabled()`.
- Simplify the console methods wrapping and move the `consoleFromVM`
  storage to `internal/util/inspector.js`

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

joyeecheung commented:
> This change messes the top stack frame - instead of displaying user code location it returns internal/util/inspector.js:42

Is it possible to write a test for this? It seems pretty fragile to have this contract depended on by external tools without any tests.

eugeneo commented:
I created a new pull request with a test: https://github.com/nodejs/node/pull/25455

joyeecheung commented:
Rebased after https://github.com/nodejs/node/pull/25455 and removed `consoleCall` changes. @eugeneo @jdalton PTAL.

CI: https://ci.nodejs.org/job/node-test-pull-request/20153/

jdalton commented on lib/internal/util/inspector.js:
```diff
@@ -20,6 +18,49 @@ function sendInspectorCommand(cb, onError) {
+  consoleAPIModule.paths =
+      CJSModule._nodeModulePaths(cwd).concat(CJSModule.globalPaths);
+  commandLineApi.require = makeRequireFunction(consoleAPIModule);
+}
```
👆 Could someone explain this bit. I'm unfamiliar with the `'<inspector console>'` and `commandLineApi.require` bits. Is there a `require` exposed in the Chrome inspector console?

devsnek replied:
The inspector console is the one in the chrome inspector. By doing this, a `require` is exposed when you connect a chrome inspector to node.

jdalton replied:
Is this something user accessible?

joyeecheung replied:
@jdalton: it's accessed when using `require` in DevTools console.

joyeecheung commented:
Ping @eugeneo I've removed the `consoleCall` changes, can you take a look again? I'd like to land this so that `'internal/modules/cjs/loader'` does not have to be loaded this early during bootstrap and can be loaded in `prepareUserCodeExecution` so that we can split the bootstrapping process from the execution mode selection which helps the snapshot effort.

joyeecheung commented:
I am going to dismiss https://github.com/nodejs/node/pull/25450#pullrequestreview-191845760 since it's been addressed a week ago without response, and it no longer applies to this PR since this passes the test added in https://github.com/nodejs/node/pull/25455

joyeecheung commented:
Fresh CI: https://ci.nodejs.org/job/node-test-pull-request/20280/

joyeecheung commented on lib/internal/bootstrap/node.js:
```diff
@@ -679,13 +679,17 @@ function setupGlobalConsole() {
-    inspector.addInspectorApis(consoleFromNode, consoleFromVM);
+    const inspector = NativeModule.require('internal/util/inspector');
     // This will be exposed by `require('inspector').console` later.
     inspector.consoleFromVM = consoleFromVM;
```
The console from the VM is only accessible during bootstrap, it gets overridden with our console implementation - an alternative may be to store it in C++ and make it accessible (and cached) through `'internal/util/inspector'`, but otherwise this has to be done in `lib/internal/bootstrap/node.js` one way or another (stored either as a value or through a setter)

BridgeAR replied:
`consoleFromVM` is passed to `inspector.wrapConsole()` in the line below. Therefore the `consoleFromVM` variable in `lib/internal/util/inspector.js` could be set during the `wrapConsole` function call.

But it's just a nit and is not blocking for me.

joyeecheung commented:
Resume: https://ci.nodejs.org/job/node-test-pull-request/20282/
